bdk_electrum_example/src/electrum.rs

```diff
@@ -0,0 +1,151 @@
+        &self,
+        scripts: BTreeMap<K, impl Iterator<Item = (u32, Script)>>,
+        stop_gap: Option<usize>,
+        local_chain: &BTreeMap<u32, BlockHash>,
+    ) -> Result<(SparseChain, BTreeMap<K, u32>)> {
+        let mut sparse_chain = SparseChain::default();
+
+        // Check for reorgs.
+        // In case of reorg, new checkpoints until the last common checkpoint is added to the structure
+        for (&existing_height, &existing_hash) in local_chain.iter().rev() {
+            let current_hash = self
+                .client
+                .block_header(existing_height as usize)?
+                .block_hash();
+            sparse_chain
+                .insert_checkpoint(BlockId {
+                    height: existing_height,
+                    hash: current_hash,
+                })
+                .expect("should not collide");
+
+            if current_hash == existing_hash {
+                break;
+            }
+        }
+
+        // Insert the new tip
+        let (tip_height, tip_hash) = self.get_tip()?;
+        sparse_chain
+            .insert_checkpoint(BlockId {
+                height: tip_height,
+                hash: tip_hash,
+            })
+            .expect("Should not collide");
+
+        let mut keychain_index_update = BTreeMap::new();
+
+        // Fetch Keychain's last_active_index and all related txids.
+        // Add them into the KeyChainScan
+        for (keychain, mut scripts) in scripts.into_iter() {
+            let mut last_active_index = 0;
+            let mut unused_script_count = 0usize;
+            let mut script_history_txid = Vec::<(Txid, TxHeight)>::new();
+
+            loop {
+                let (index, script) = scripts.next().expect("its an infinite iterator");
+
+                let history = self
+                    .script_get_history(&script)?
+                    .iter()
+                    .map(|history_result| {
+                        if history_result.height > 0 {
+                            return (
+                                history_result.tx_hash,
+                                TxHeight::Confirmed(history_result.height as u32),
+                            );
+                        } else {
+                            return (history_result.tx_hash, TxHeight::Unconfirmed);
+                        };
+                    })
+                    .collect::<Vec<(Txid, TxHeight)>>();
+
+                if history.is_empty() {
+                    unused_script_count += 1;
+                } else {
+                    last_active_index = index;
+                    script_history_txid.extend(history.iter());
+                    unused_script_count = 0;
+                }
+
+                if unused_script_count >= stop_gap.unwrap_or(usize::MAX) {
+                    break;
+                }
+            }
+
+            for (txid, index) in script_history_txid {
+                sparse_chain.insert_tx(txid, index)?;
```

Comment:
Deal with this error here.

Reply:
You are throwing the error on `insert_checkpoint` so doesn't this just return with an error if there was a re-org (since `insert_checkpoint` returns an error if the hashes don't match). It seems simpler just to check that the hash at the start tip height is the same and throw an error if not.

It looks like you are trying to do your best even in the face of a re-org and still return a sparse chain. This is a cool idea. It might require some more thought though. For now I suggest just returning a `ReOrg` error if there was a re-org while you are doing the scan and we can make an issue about how to do better than this.

Inserting some extra checkpoints is ok -- we do that in esplora but in that case we have a single call to fetch the most recent ones. It might not be worth it here.